contracts/contracts/harvest/Harvester.sol

```diff
@@ -0,0 +1,384 @@
+     */
+    function _harvest(address _strategyAddr) internal {
+        IStrategy strategy = IStrategy(_strategyAddr);
+        address[] memory rewardTokenAddresses = strategy
+            .getRewardTokenAddresses();
+
+        if (rewardTokenAddresses.length > 0) {
+            strategy.collectRewardTokens();
+        }
+    }
+
+    /**
+     * @dev Swap all supported swap tokens for stablecoins via Uniswap.
+     */
+    function _swap() internal {
+        address[] memory allStrategies = IVault(vaultAddress)
+            .getAllStrategies();
+
+        for (uint256 i = 0; i < swapTokens.length; i++) {
+            _swap(swapTokens[i]);
+        }
+    }
+
+    /**
+     * @dev Swap a record token for stablecoins for Uniswap. The token must have
+     *       a registered price feed with the price provider.
+     * @param _swapToken Address of the token to swap.
+     */
+    function _swap(address _swapToken)
+        internal
+        returns (uint256[] memory swapResult)
+    {
+        RewardTokenConfig memory tokenConfig = rewardTokenConfigs[_swapToken];
+        require(
+            tokenConfig.uniswapV2CompatibleAddr != address(0),
+            "Swap token is missing token configuration."
+        );
+
+        address priceProvider = IVault(vaultAddress).priceProvider();
+
+        IERC20 swapToken = IERC20(_swapToken);
+        uint256 balance = swapToken.balanceOf(address(this));
+        if (balance > 0) {
+            uint256 maxBalanceToSwap = balance;
+            if (tokenConfig.liquidationLimit != 0) {
+                maxBalanceToSwap = Math.min(
+                    balance,
+                    tokenConfig.liquidationLimit
+                );
+            }
+
+            // This'll revert if there is no price feed
+            uint256 oraclePrice = IOracle(priceProvider).price(_swapToken);
+            // Oracle price is 1e8, USDT output is 1e6
+            uint256 minExpected = (maxBalanceToSwap *
+                oraclePrice *
+                (1e4 - tokenConfig.allowedSlippageBps)).scaleBy( // max allowed slippage
+                6,
+                Helpers.getDecimals(_swapToken) + 8
+            ) / 1e4; // fix the max slippage decimal position
+
+            // Uniswap redemption path
+            address[] memory path = new address[](3);
+            path[0] = _swapToken;
+            path[1] = IUniswapV2Router(tokenConfig.uniswapV2CompatibleAddr)
+                .WETH();
+            path[2] = usdtAddress;
+
+            swapResult = IUniswapV2Router(tokenConfig.uniswapV2CompatibleAddr)
+                .swapExactTokensForTokens(
+                    maxBalanceToSwap,
+                    minExpected,
+                    path,
+                    address(this),
+                    block.timestamp
+                );
+
+            IERC20 usdt = IERC20(usdtAddress);
+            uint256 usdtBalance = usdt.balanceOf(address(this));
+            uint16 vaultBps = 1e4 - tokenConfig.harvestRewardBps;
```

Comment:
This roughly does:

```
a = pct * total / decimals;
b = (1-pct) * total / decimals;
```

A "standard" way to calculate a percentage split of something has evolved in solidity, which looks like this:

```python
a = pct * total / decimals;
b = total - a;
```

This has a few advantages:

- No rounding errors. The other method will have rounding errors on both A and B, which will usualy result in a + b not quite equaling the total.
- Hard math in just one place. Getting all the decimals right is a pain. With the standard way, you only have to do this once.
- Can't screw up the amounts beyond the total when using uints. No matter what you do, `a + b` is always going to exactly equal total. If `a` is miscalculated too high, the subtraction will throw and save you. If A is too low, well, B gets more more, but no funds have gone into limbo, and B can't get more than total.

Reply:
Thanks this makes so much more sense. I was researching this and I think I read that divisions in solidity always round down. So the 2 numbers should never be above total, but there are as you say still small rounding errors. The suggested fix is so much cleaner.